src/iceberg/avro/avro_schema_util.cc

```diff
@@ -783,4 +785,201 @@ Result<SchemaProjection> Project(const Schema& expected_schema,
+    case ::avro::AVRO_ARRAY:
+      return CreateArrayNodeWithFieldIds(original_node, name_mapping);
+    case ::avro::AVRO_MAP:
+      return CreateMapNodeWithFieldIds(original_node, name_mapping);
+    case ::avro::AVRO_UNION:
+      return CreateUnionNodeWithFieldIds(original_node, name_mapping);
+    case ::avro::AVRO_BOOL:
+    case ::avro::AVRO_INT:
+    case ::avro::AVRO_LONG:
+    case ::avro::AVRO_FLOAT:
+    case ::avro::AVRO_DOUBLE:
+    case ::avro::AVRO_STRING:
+    case ::avro::AVRO_BYTES:
+    case ::avro::AVRO_FIXED:
+      // For primitive types, just return a copy
+      return original_node;
+    case ::avro::AVRO_NULL:
+    case ::avro::AVRO_ENUM:
+    default:
+      return InvalidSchema("Unsupported Avro type for field ID application: {}",
+                           ToString(original_node));
+  }
+}
+
+Result<::avro::NodePtr> CreateRecordNodeWithFieldIds(const ::avro::NodePtr& original_node,
+                                                     const NameMapping& name_mapping) {
+  auto new_record_node = std::make_shared<::avro::NodeRecord>();
+  new_record_node->setName(original_node->name());
+
+  for (size_t i = 0; i < original_node->leaves(); ++i) {
+    const std::string& field_name = original_node->nameAt(i);
+    ::avro::NodePtr field_node = original_node->leafAt(i);
+
+    // Try to find field ID by name in the name mapping
+    if (auto field_ref = name_mapping.Find(field_name)) {
+      if (field_ref->get().field_id.has_value()) {
+        // Add field ID attribute to the new node
+        ::avro::CustomAttributes attributes;
+        attributes.addAttribute(std::string(kFieldIdProp),
+                                std::to_string(field_ref->get().field_id.value()), false);
+        new_record_node->addCustomAttributesForField(attributes);
+      }
+
+      // Recursively apply field IDs to nested fields if they exist
+      if (field_ref->get().nested_mapping && field_node->type() == ::avro::AVRO_RECORD) {
+        const auto& nested_mapping = field_ref->get().nested_mapping;
+        auto fields_span = nested_mapping->fields();
+        std::vector<MappedField> fields_vector(fields_span.begin(), fields_span.end());
+        auto nested_name_mapping = NameMapping::Make(std::move(fields_vector));
+
+        ICEBERG_ASSIGN_OR_RAISE(
+            auto new_nested_node,
+            CreateAvroNodeWithFieldIds(field_node, *nested_name_mapping));
+        new_record_node->addName(field_name);
+        new_record_node->addLeaf(new_nested_node);
+      } else {
+        // Recursively apply field IDs to child nodes
+        ICEBERG_ASSIGN_OR_RAISE(auto new_field_node,
+                                CreateAvroNodeWithFieldIds(field_node, name_mapping));
+        new_record_node->addName(field_name);
+        new_record_node->addLeaf(new_field_node);
+      }
+    } else {
+      // Recursively apply field IDs to child nodes even if no mapping found
+      ICEBERG_ASSIGN_OR_RAISE(auto new_field_node,
```

Comment:
Should we return error status if field_id is missing? Then we don't need to validate this again in the Avro reader after applying name mapping.